src/coreclr/src/gc/gcpriv.h

```diff
@@ -693,32 +694,62 @@ class alloc_list
 
 class allocator
 {
-    size_t num_buckets;
-    size_t frst_bucket_size;
+    int first_bucket_bits;
+    unsigned int num_buckets;
     alloc_list first_bucket;
     alloc_list* buckets;
     alloc_list& alloc_list_of (unsigned int bn);
     size_t& alloc_list_damage_count_of (unsigned int bn);
 
 public:
-    allocator (unsigned int num_b, size_t fbs, alloc_list* b);
+    allocator (unsigned int num_b, int fbb, alloc_list* b);
+
     allocator()
     {
         num_buckets = 1;
-        frst_bucket_size = SIZE_T_MAX;
+        first_bucket_bits = sizeof(size_t) * 8;
+    }
+
+    unsigned int number_of_buckets()
+    {
+        return num_buckets;
+    }
+
+    // skip buckets that cannot possibly fit "size" and return the next one
+    // there is always such bucket since the last one fits everything
+    unsigned int first_suitable_bucket(size_t size)
```

Comment:
>unsigned int first_suitable_bucket(size_t size) [](start = 4, length = 47)

have you actually observed any perf difference doing this? I doubt it.

Reply:
On a microbenchmark doing a lot of POH allocations this becomes noticeable. I saw about 10% improvement from these changes.

The hot spots that I saw: taking msl, iterating through buckets, splitting a free chunk in two (when chunk is big, makefreelistmay cause cache misses) 
The others are hard to avoid or improve though. This one was fairly easy.

Reply:
can you share your benchmark? what sort of perf changes were done on this in general? are you running the default perf scenarios with the infra? this new scenario should be added to gcperfsim.

---
In reply to: [399581040](https://github.com/dotnet/runtime/pull/34215#discussion_r399581040) [](ancestors = 399581040)

Reply:
An example:

```cs    
    private static void AllocateUninitializedPinned()
    {
        System.Console.Write("AllocateUninitializedPinned: ");
        var r = new Random(1234);
        Stopwatch sw = Stopwatch.StartNew();
        for (int i = 0; i < iters; i++)
        {
            int size = r.Next(2, 10000);
            var arr = GC.AllocateUninitializedArray<long>(size, pinned: true);
            Add(arr);   // noop when not retaining, store in a static list when do
        }

        sw.Stop();
        System.Console.WriteLine(sw.ElapsedMilliseconds);
    }
```

I had a variety of these - with not retaining arrays, retaining until after the iteration is over, retaining randomly in an array

like:
```cs
    private static void Add(long[] arr)
    {
        if (r.Next(100) > 50)
            staticArray[arr.Length & 2047] = arr;
    }
```

The not retaining anything scenario was the most sensitive to the cost of the allocation codepath.

Reply:
I ran the standard infra with these chenges, but I mostly see noise. If there are improvements, they are too small in the scenarios measured.

Yes, we should add POH scenarios to infra. At least for regression testing.
We may need to settle on what would be the scenario too look at - how big allocations are. How much is retained, etc. 
If we track scenarios for LOH, POH scenarios could be very similar, except for size.

Reply:
LOH and POH are not necessarily that similar - LOH is compacted by default for hardlimit and POH can't be compacted; LOH has in general much larger object sizes. let's sync up next week on what scenarios we want to test.

Reply:
Yes. For regression testing we need specific scenarios.

The tweaks I did here were just for something that stood out from looking at the code and VTune.